autode/opt/coordinates/base.py

```diff
@@ -0,0 +1,291 @@
+    """Coordinates used to perform optimisations"""
+
+    implemented_units = [ang, nm, pm, m]
+
+    def __new__(cls,
+                input_array: Union[Sequence, np.ndarray],
+                units:       Union[str, 'autode.units.Unit']
+                ) -> 'OptCoordinates':
+        """New instance of these coordinates"""
+
+        arr = super().__new__(cls, np.array(input_array), units)
+
+        arr._e = None              # Energy
+        arr._g = None              # Gradient: dE/dX
+        arr._h = None              # Hessian:  d2E/dX_idX_j
+        arr._h_inv = None          # Inverse of the Hessian: H^-1
+        arr.B = None               # Wilson B matrix
+        arr.B_T_inv = None         # Generalised inverse of B
+        arr.U = np.eye(len(arr))   # Transform matrix
+
+        return arr
+
+    def __array_finalize__(self, obj: 'OptCoordinates') -> None:
+        """See https://numpy.org/doc/stable/user/basics.subclassing.html"""
+
+        for attr in ('units', '_e', '_g', '_h', '_h_inv', 'U', 'B', 'B_T_inv'):
+            self.__dict__[attr] = getattr(obj, attr, None)
+
+        return None
+
+    @property
+    def raw(self) -> np.ndarray:
+        """Raw numpy array of these coordinates"""
+        return np.array(self, copy=True)
+
+    @property
+    def e(self) -> Optional[PotentialEnergy]:
+        """
+        Energy
+
+        -----------------------------------------------------------------------
+        Returns:
+            (PotentialEnergy | None): E
+        """
+        return self._e
+
+    @e.setter
+    def e(self, value):
+        """Set the energy"""
+        self._e = None if value is None else PotentialEnergy(value)
+
+    @property
+    def g(self) -> Optional[np.ndarray]:
+        r"""
+        Gradient of the energy
+
+        .. math::
+            G = \nabla E
+                \equiv
+                \left\{\frac{\partial E}{\partial\boldsymbol{R}_{i}}\right\}
+
+        where :math:`\boldsymbol{R}` are a general vector of coordinates.
+
+        -----------------------------------------------------------------------
+        Returns:
+            (np.ndarray | None): G
+        """
+        return self._g
+
+    @g.setter
+    def g(self, value: np.ndarray):
+        """Set the gradient of the energy"""
+        self._g = value
+
+    @property
+    def h(self) -> Optional[np.ndarray]:
+        r"""
+        Hessian (second derivative) matrix of the energy
+
+        .. math::
+            H = \begin{pmatrix}
+                \frac{\partial^2 E}
+                     {\partial\boldsymbol{R}_{0}\partial\boldsymbol{R}_{0}}
+                & \cdots \\
+                \vdots & \ddots
+                \end{pmatrix}
+
+        where :math:`\boldsymbol{R}` are a general vector of coordinates.
+
+        -----------------------------------------------------------------------
+        Returns:
+            (np.ndarray | None): H
+        """
+        if self._h is None and self._h_inv is not None:
```

Comment:
Is there a way you could check this once instead of each time the `h` property is accessed? If this condition doesn't change state regularly...

Reply:
`h` is updated every optimisaiton step so it changes state pretty reguarly – I think it's safest to leave this check, and performance isn't super important